File: src/commands/docs.ts
import { type Renderable, detailDoc, encode } from "../output/index.js";
import type { DocsResult } from "../output/index.js";
import { searchDocs } from "../search.js";
import {
  DEFAULT_QUERY_LIMIT,
  defineGlobalCommand,
  loadBundle,
  parseLimit,
  resolveFormat,
} from "./shared.js";

export const docsCommand = defineGlobalCommand({
  meta: { name: "docs", description: "Search documentation and guideline pages" },
  args: {
    term: { type: "positional", description: "search term", required: true },
    limit: {
      type: "string",
      description: "maximum results (0 = all)",
      default: String(DEFAULT_QUERY_LIMIT),
    },
  },
  async run({ args }) {
    const format = resolveFormat(args);
    const bundle = await loadBundle(args);

    const term = args.term;
    const limit = parseLimit(args.limit);

    const result: DocsResult = {
      term,
      docs: searchDocs(bundle.docs.docs ?? {}, term, limit).map((m) => detailDoc(m.doc)),
    };
    if (bundle.warnings.length > 0) result.warnings = bundle.warnings;

    const payload: Renderable = { kind: "docs", value: result };
    process.stdout.write(encode(format, payload));
  },
});

File: internal/cli/docs.go
package cli

import (
	"os"

	"github.com/spf13/cobra"

	"github.com/adriankarlen/storyquery/internal/output"
	"github.com/adriankarlen/storyquery/internal/search"
)

func newDocsCmd(gf *globalFlags) *cobra.Command {
	var limit int
	cmd := &cobra.Command{
		Use:   "docs <term>",
		Short: "Search documentation and guideline pages",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := resolveFormat(gf)
			if err != nil {
				return err
			}
			bundle, err := loadBundle(cmd.Context(), gf)
			if err != nil {
				return err
			}

			term := args[0]
			matches := search.Docs(bundle.Docs.Docs, term, limit)
			res := output.DocsResult{Term: term, Warnings: bundle.Warnings}
			for _, m := range matches {
				res.Docs = append(res.Docs, output.DetailDoc(m.Doc))
			}
			return output.Encode(os.Stdout, format, res)
		},
	}
	cmd.Flags().IntVar(&limit, "limit", defaultQueryLimit, "maximum results (0 = all)")
	return cmd
}
